arborist/server.go

```diff
@@ -1304,13 +1304,49 @@ func (server *Server) handleUserRevokeAll(w http.ResponseWriter, r *http.Request
 func (server *Server) handleUserRevokePolicy(w http.ResponseWriter, r *http.Request) {
 	username := mux.Vars(r)["username"]
 	policyName := mux.Vars(r)["policyName"]
-	errResponse := revokeUserPolicy(server.db, username, policyName, getAuthZProvider(r))
-	if errResponse != nil {
-		errResponse.log.write(server.logger)
-		_ = errResponse.write(w, r)
+	authzProvider := getAuthZProvider(r)
+	policyInfo, err := fetchUserPolicyInfo(server.db, username, policyName)
+
+	if err != nil {
+		server.logger.Info("Error Fetching policy Info: %s", err.Error())
+		msg := fmt.Sprintf("Error Fetching policy Info: %s", err.Error())
+		response := newErrorResponse(msg, http.StatusInternalServerError, nil)
+		_ = response.write(w, r)
 		return
 	}
-	server.logger.Info("revoked policy %s for user %s", policyName, username)
+
+	if policyInfo != nil {
+
+		dbAuthzProvider := ""
+		providerExists := policyInfo.AuthzProvider.Valid
+		if providerExists {
+			dbAuthzProvider = policyInfo.AuthzProvider.String
+		}
+		server.logger.Debug("Policy - {name: %s, authz_provider: %s, expires_at: %s} assigned to user %s",
+			policyInfo.PolicyName, dbAuthzProvider, policyInfo.ExpiresAt, policyInfo.Username)
+
+		if !authzProvider.Valid || (providerExists && dbAuthzProvider == authzProvider.String) {
+			errResponse := revokeUserPolicy(
+				server.db, username, policyName, authzProvider)
+			if errResponse != nil {
+				errResponse.log.write(server.logger)
+				_ = errResponse.write(w, r)
+				return
+			}
+			server.logger.Info("revoked policy %s for user %s", policyName, username)
+		} else {
+			server.logger.Info("Cannot revoke policy `%s`. Authz_provider - `%s` and `%s` Mismatch",
+				policyName, policyInfo.AuthzProvider.String, authzProvider.String)
+			msg := fmt.Sprintf("Cannot revoke policy `%s`. Authz_provider Mismatch", policyName)
+			errResponse := newErrorResponse(msg, http.StatusUnauthorized, nil)
+			errResponse.log.write(server.logger)
+			_ = errResponse.write(w, r)
+			return
+		}
+	} else {
+		server.logger.Error("Policy `%s` does not exist for user `%s`. Check if it is assigned through a group.",
```

Comment:
from the requirements:

> Revoking a policy the user doesn't have => nothing changed but the action was successful, so we can keep returning 204 (maybe add a debug log?)
Revoking a policy that was granted via a group instead of directly to the user => Add logic to return a 400 error (the policy should be removed from the group instead)

with this implementation i think we have to choose either 204 or 400 for both cases. Right now i think you return 204 for both, but 400 would make more sense to me, so the user doesn't think something was revoked when it wasn't (group case). And 400 for trying to delete something that isn't there sounds ok. Wdyt?

Reply:
Yeah. We can say that. It will be in a way giving more info to the user than we used to. Say, a particular policy doesn't exist for the current user. If that's not a problem, I don't see why returning a `Bad Request` would be any issue. I'm gonna go ahead and update it. Let me know if you've got any more thoughts on this.